### ISSUE TITLE
Question: check method for external (python) module

### ISSUE THREAD

**Issue by timoles** [rapid7/metasploit-framework#12471]
As in the title:
Is there a way to implement a [check()](https://github.com/rapid7/metasploit-framework/wiki/How-to-write-a-check()-method) for external python modules?
I checked the documentation and skimmed through the external module [implementation](https://github.com/rapid7/metasploit-framework/tree/master/lib/msf/core/modules/external/python), but didn't find any info about it.
Thanks in advance :)

**Comment by bwatters-r7** [comment]
@acammack-r7, cough, cough

**Comment by acammack-r7** [comment]
Hey @timoles, you will need to pass an addition callback to [`module.run`](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/modules/external/python/metasploit/module.py#L84) that returns a check code status (one of `safe`, `detected`, `appears`, `vulnerable`, or `unknown`). The only external module to have a check right now is [exploit/linux/smtp/haraka](https://github.com/rapid7/metasploit-framework/blob/master/modules/exploits/linux/smtp/haraka.py#L112), so if there is something missing in the implementation you need please let us know! I'll also be sure to add this to the docs.

**Comment by timoles** [comment]
Thank you very much @acammack-r7  for the help. 
With the haraka example I got the check working on an external module with the type "remote_exploit_cmd_stager". 
I have one question though: Am I right in the assumption that there currently is no "check" command for external scanner modules? I tried to figure it out through the external module source code, but I'm not quite sure.

---

My current assumption is that all type templates which implement [this functionality/template](https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/modules/external/templates/common_check.erb) also provide a "check()" function, but according to the "ruby" [docs](https://github.com/rapid7/metasploit-framework/wiki/How-to-write-a-check()-method#remote-check-example) it should also work with normal scanners?

Any help is appreciated, and thank you for the help already provided :)

Ps.: To make the problem more clear, here are a few code snippets of what I tried
I made a minimalistic exploit and scanner with following "body"
```python
def run(args):
    module.log('Module dependency (requests) is missing, cannot continue', level="error")
def check_host(args):
    module.log("Exploit HTTP response code: {}".format("Test"), level="good")
    return "detected"
if __name__ == '__main__':
    module.run(metadata, run, soft_check=check_host)
```
Following Metadata bring the results:
`metadata = ... 'type': 'remote_exploit_cmd_stager', ...` => check() function works as expected
`metadata = ... 'type': 'single_host_login_scanner',#single_scanner ` => Error message: "check function not implemented"

**Comment by acammack-r7** [comment]
Good catch, you will also need to add this line to the login scanner template: https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/modules/external/templates/remote_exploit_cmd_stager.erb#L34

**Comment by ccondon-r7** [comment]
:teamwork-emoji: Closing this, feel free to ping us again or reach out on Slack if you have more questions! Thanks, for the knowledge-droppin', Adam.

**Comment by acammack-r7** [comment]
Re-opening as the template in question still does not support `check`.
